<center><h1>Inteligencia Artificial</header1></center>

Presentado por: María José Párraga Moreira  <br>
Fecha: 27/7/2022

# Importación de librerias necesarias

In [307]:
#Para esta actividad se importarán las siguientes librerías:
#Para esta actividad se importarán las siguientes librerías:
#importar libreria pandas
import pandas as pd

## Cargar el Dataset

In [308]:
#Código para cargar el Dataset
#nombre del Dataset
data = 'gender_classification_v7.csv'
#Transfromando el ; del dataset 
data2 = pd.read_csv(data, header=0, sep=',')

#Creando un nuevo dataset
df = pd.DataFrame(data2)
#imprimir los datos del dataset
df.head()

long_hair  forehead_width_cm  forehead_height_cm  nose_wide  nose_long  \
0          1               11.8                 6.1          1          0   
1          0               14.0                 5.4          0          0   
2          0               11.8                 6.3          1          1   
3          0               14.4                 6.1          0          1   
4          1               13.5                 5.9          0          0   

   lips_thin  distance_nose_to_lip_long  gender  
0          1                          1    Male  
1          1                          0  Female  
2          1                          1    Male  
3          1                          1    Male  
4          0                          0  Female

## Descripción de la fuente del Dataset

El presente set de datos lo que busca es llevar a cabo una clasificicon sobre el genero de las personas en base a diferentes atributos o valores de entrada como el largo del cabello, ancho y altura de la frente, datos de la nariz, distancia entre la nariz y los labios.

## Explique el problema a resolver. 
Con el fin de poder obtener una correcta clasificacion del genero de una persona se utilizara el metodo SVM para dicha clasificacion en base a 7 diferentes atributos o valores de entrada. Esto brinda la facilidad de tener una estandarizacion del genero de las personas en base a los atributos antes mencionados.
Variables de Entrada:
long_hair->categorico este valor tiene 0 si el cabello no es largo y 1 si l es	
forehead_width_cm	-> continua valor en cm hace referencia al ancho de la frente
forehead_height_cm	-> continua valor en cm que hace referencia a la altura de la frente
nose_wide->categorica entrega valor de 0 si la nariz no es amplia y 1 si lo es
nose_long	-> categorica entrega valor de 1 si la nariz es largo y 0 sino
lips_thin	->categoirica con valor de 1 si son labios delgados y 0 sino
distance_nose_to_lip_long-> categoirica indica 1 si la distancia es larga entre nariz y labios 0 sino

Variable Objetivo:
Genero-> categorica entrega genero masculino o femenino en base al genero de la persona


## Caracterización del Dataset

Realice una descripción de los datos con:

>- Número de instancias en total.
>- Número de atributos de entrada, su significado y tipo.
>- Estadísticas de la variable objetivo.
>- Estadísticas los atributos en relación con la variable objetivo.


In [309]:
#Código que responde a la descripción anterior
df.describe()

long_hair  forehead_width_cm  forehead_height_cm    nose_wide  \
count  5001.000000        5001.000000         5001.000000  5001.000000   
mean      0.869626          13.181484            5.946311     0.493901   
std       0.336748           1.107128            0.541268     0.500013   
min       0.000000          11.400000            5.100000     0.000000   
25%       1.000000          12.200000            5.500000     0.000000   
50%       1.000000          13.100000            5.900000     0.000000   
75%       1.000000          14.000000            6.400000     1.000000   
max       1.000000          15.500000            7.100000     1.000000   

         nose_long    lips_thin  distance_nose_to_lip_long  
count  5001.000000  5001.000000                5001.000000  
mean      0.507898     0.493101                   0.498900  
std       0.499988     0.500002                   0.500049  
min       0.000000     0.000000                   0.000000  
25%       0.000000     0.000000                   0.000000  
50%       1.000000     0.000000                   0.000000  
75%       1.000000     1.000000                   1.000000  
max       1.000000     1.000000                   1.000000

Se puede observar en el presente set que se dispone de 7 valores de entrada y de 1 variable objetivo que es la que se comento en la seccion anterior. El set de datos cuenta con un total de 5001 datos en bruto que no han sido trabajado. Para las variables continuas como forehead_width se tiene una media de 13.18 cm y en cuato forehead_height de 5.96. Se tiene en su mayoria vriables categoricas siendo las 2 ultimas aquellas continuas.

In [310]:
#Estadistica de variable obejtivo
gender = df.gender
gender.describe()

count       5001
unique         2
top       Female
freq        2501
Name: gender, dtype: object

In [311]:
#Estadísticas los atributos en relación con la variable objetivo.
y = df[['long_hair','forehead_width_cm','forehead_height_cm','nose_wide','nose_long','lips_thin','distance_nose_to_lip_long']]
#mostrar la información
y.describe()

long_hair  forehead_width_cm  forehead_height_cm    nose_wide  \
count  5001.000000        5001.000000         5001.000000  5001.000000   
mean      0.869626          13.181484            5.946311     0.493901   
std       0.336748           1.107128            0.541268     0.500013   
min       0.000000          11.400000            5.100000     0.000000   
25%       1.000000          12.200000            5.500000     0.000000   
50%       1.000000          13.100000            5.900000     0.000000   
75%       1.000000          14.000000            6.400000     1.000000   
max       1.000000          15.500000            7.100000     1.000000   

         nose_long    lips_thin  distance_nose_to_lip_long  
count  5001.000000  5001.000000                5001.000000  
mean      0.507898     0.493101                   0.498900  
std       0.499988     0.500002                   0.500049  
min       0.000000     0.000000                   0.000000  
25%       0.000000     0.000000                   0.000000  
50%       1.000000     0.000000                   0.000000  
75%       1.000000     1.000000                   1.000000  
max       1.000000     1.000000                   1.000000

## Preprocesamiento del dataset. Transformaciones previas necesarias para la modelación

In [312]:
y = df[['gender']]
#Separando las caracteristicas 
X = df[['long_hair','forehead_width_cm','forehead_height_cm','nose_wide','nose_long','lips_thin','distance_nose_to_lip_long']]

In [313]:
#Comprobación de valores nulos en datos de entrada
X.isnull().sum()

long_hair                    0
forehead_width_cm            0
forehead_height_cm           0
nose_wide                    0
nose_long                    0
lips_thin                    0
distance_nose_to_lip_long    0
dtype: int64

In [314]:
#Comprobación de valores nullos de objetivo
y.isnull().sum()

gender    0
dtype: int64

In [315]:
### Estandarización de la data ###
# importar metodo de scalado standar
from sklearn.preprocessing import StandardScaler
# Importación de método para dividir datos de prueba 
from sklearn.model_selection import train_test_split
# Formar objeto de scala para entradas
VariablesEntradaScaler=StandardScaler()
# Formar objeto de escala para objetivo
VariableObjetivoScaler=StandardScaler()

In [316]:
#Almacenamiento del objeto de ajuste para referencia posterior
#Datos de entrada
VariablesEntradaFit=VariablesEntradaScaler.fit(X)

In [317]:
#Generación de los valores estandarizados de X e y
#Datos e entrada
X=VariablesEntradaFit.transform(X)

## División del dataset en datos de entrenamiento y datos de test 

In [318]:
#Código que realice la división en entrenamiento y test, de acuerdo con la estretgia de evaluación planeada. Describa cuál es.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

## Modelamiento

In [319]:
#Códo del modelo
#Importo Libreria del Modelo
from sklearn import svm
#Defino el Modelo
clf = svm.SVC(kernel='linear')


In [320]:
#Entrenamiento
clf.fit(X_train, y_train)

c:\users\usuario\appdata\local\programs\python\python38\lib\site-packages\sklearn\utils\validation.py:985: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(kernel='linear')

In [321]:
#Test
y_pred = clf.predict(X_test)

## Evaluación del Modelo

Construya un o dos párrafos con los principales hallazgos. Cómo está funcionando el modelo? Calidad en los resultados de predicción. 

In [322]:
### Evaluando el modelo
#Libreria para la Evaluacion
from sklearn import metrics

#Determinar cuan correcto es el modelo
print("Accuracy:", metrics.accuracy_score(y_test, y_pred))


Accuracy: 0.9680213191205863


In [323]:
#Libreria para reporte de resultados
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

      Female       0.97      0.97      0.97       745
        Male       0.97      0.97      0.97       756

    accuracy                           0.97      1501
   macro avg       0.97      0.97      0.97      1501
weighted avg       0.97      0.97      0.97      1501

